x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/ErrorsForCasesWithoutExamplesTestCase.java

```diff
@@ -0,0 +1,167 @@
+            assertTrue("expected unresolved " + expression, expression.typeResolved().unresolved());
+            assertThat(expression.typeResolved().message(), expectedTypeErrorMatcher(validPerPosition, signature));
+            checked++;
+        }
+        logger.info("checked {} signatures", checked);
+        assertThat("didn't check any signatures", checked, greaterThan(0));
+    }
+
+    private Stream<List<DataType>> missingSignatures(List<TestCaseSupplier> cases, Set<List<DataType>> valid) {
+        return cases.stream()
+            .map(s -> s.types().size())
+            .collect(Collectors.toSet())
+            .stream()
+            .flatMap(AbstractFunctionTestCase::allPermutations)
+            .filter(types -> valid.contains(types) == false)
+            /*
+             * Skip any cases with more than one null. Our tests don't generate
+             * the full combinatorial explosions of all nulls - just a single null.
+             * Hopefully <null>, <null> cases will function the same as <null>, <valid>
+             * cases.
+             */
+            .filter(types -> types.stream().filter(t -> t == DataType.NULL).count() <= 1);
+    }
+
+    protected static String sourceForSignature(List<DataType> signature) {
+        StringBuilder source = new StringBuilder();
+        for (DataType type : signature) {
+            if (false == source.isEmpty()) {
+                source.append(", ");
+            }
+            source.append(type.typeName());
+        }
+        return source.toString();
+    }
+
+    /**
+     * Build the expected error message for an invalid type signature.
+     */
+    protected static String typeErrorMessage(
+        boolean includeOrdinal,
+        List<Set<DataType>> validPerPosition,
+        List<DataType> signature,
+        AbstractFunctionTestCase.PositionalErrorMessageSupplier expectedTypeSupplier
+    ) {
+        int badArgPosition = -1;
+        for (int i = 0; i < signature.size(); i++) {
+            if (validPerPosition.get(i).contains(signature.get(i)) == false) {
+                badArgPosition = i;
+                break;
+            }
+        }
+        if (badArgPosition == -1) {
+            throw new IllegalStateException(
+                "Can't generate error message for these types, you probably need a custom error message function"
+            );
+        }
+        String ordinal = includeOrdinal ? TypeResolutions.ParamOrdinal.fromIndex(badArgPosition).name().toLowerCase(Locale.ROOT) + " " : "";
+        String source = sourceForSignature(signature);
+        String expectedTypeString = expectedTypeSupplier.apply(validPerPosition.get(badArgPosition), badArgPosition);
+        String name = signature.get(badArgPosition).typeName();
+        return ordinal + "argument of [" + source + "] must be [" + expectedTypeString + "], found value [] type [" + name + "]";
+    }
+
+    protected static String errorMessageStringForBinaryOperators(
```

Comment:
This is moved from `AbstractFunctionTestCase` and updated for the slightly different error messages this functions builds. the `+ source +` bit.